### ISSUE TITLE
[Umbrella] Skip Score plugins when coupled PreScore plugin returns Skip status

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#115745]
### What would you like to be added?

It's much similar to https://github.com/kubernetes/kubernetes/issues/114399.
In https://github.com/kubernetes/kubernetes/issues/114827, we support Skip in PreScore/Score as well. (Thanks @kidddddddddddddddddddddd)
So, the plugin can return Skip in PreScore when it doesn't do anything in Score in that scheduling cycle.

For example, if the pod doesn't have any preferred inter-pod affinity, the inter-pod affinity plugin PreScore can return Skip, and the scheduler won't run the inter-pod affinity plugin Score.

Let's change our in-tree Score plugins to return Skip.

Here, I just listed all score plugins, **not all of them may not need to return Skip in PreScore though**. Assignee, please make sure your PreScore/Score plugin really needs to return Skip or not.

Please feel free to assign what you'd like to work on. And **please comment in this issue before starting to work on so that we can avoid many people from working on the same thing!**

- [x] ImageLocality @tangwz  no need.
- [x] InterPodAffinity @utam0k https://github.com/kubernetes/kubernetes/pull/117794
- [x] NodeAffinity @sanposhiho https://github.com/kubernetes/kubernetes/pull/117024
- [x] Fit @tangwz no need.
- [x] BalancedAllocation @tangwz no need.
- [x] PodTopologySpread @utam0k https://github.com/kubernetes/kubernetes/pull/118608
- [x] SelectorSpread; no need. It'll soon be removed. https://github.com/kubernetes/kubernetes/issues/108444
- [x] TaintToleration @tangwz no need. https://github.com/kubernetes/kubernetes/pull/118324
- [x] VolumeBinding @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/115768

### Why is this needed?

The scheduler can know which plugins it really needs/doesn't need to run to get the scheduling result.
It'd contribute to the scheduling performance.

/sig scheduling
/assign

**Comment by sanposhiho** [comment]
I'm thinking work on InterPodAffinity and NodeAffinity because I worked on the Skip implementation in PreFilter/Filter for them.

**Comment by sivchari** [comment]
@sanposhiho
Please assign me to `ImageLocality`,  `PodTopologySpread` and `TaintToleration` 👍

**Comment by kidddddddddddddddddddddd** [comment]
I'd like to work on `VolumeBinding` since I worked on it before.

**Comment by tangwz** [comment]
@sanposhiho I'd like to work on `Fit`, `BalancedAllocation`.

**Comment by tangwz** [comment]
@sanposhiho @alculquicondor For `Fit` and `BalancedAllocation`, I think If the plugin's args `Resources` not set, we can return a error in PreScore Plugin, but `Skip `does not need to implement.

**Comment by alculquicondor** [comment]
Today, if `args.resources` is empty, scores return zero, but it doesn't produce an error. This doesn't make sense, but someone could have made a mistake. If we start returning an error, they would have to change their configuration to keep the scheduler running. I don't think we should be doing that.
In theory, if all nodes return the same score, we could return Skip, but since this is not a valid use case, I think we are adding complexity for no value. So I think we should leave it as is. At most, I would just produce a log line saying that not setting resources would not produce a meaningful score.

**Comment by sanposhiho** [comment]
@sivchari Do you still plan to work on it? or want anyone to take over them?

**Comment by sivchari** [comment]
@sanposhiho 
Sorry, I'll work on it.

**Comment by sivchari** [comment]
@alculquicondor 
There is no progress on the PodTopologySpread 🙇

**Comment by utam0k** [comment]
I talked to @sanposhiho elsewhere and was given a chance to work on PodAffnitiy.

**Comment by utam0k** [comment]
@sanposhiho May I ask you to assign me to this issue to avoid missing 🙏 ?

**Comment by utam0k** [comment]
@sanposhiho Is there anything else I can help with that you would like to do? Of course, I can take another issue.

**Comment by sanposhiho** [comment]
I think you can help some @sivchari has, they all have been pending for quite long time.
How about `PodTopologySpread`?

**Comment by utam0k** [comment]
@sanposhiho I can take it. Please assign me 🙏

**Comment by sanposhiho** [comment]
Can anyone be a volunteer for `TaintToleration` or `ImageLocality`?

**Comment by tangwz** [comment]
@sanposhiho assign to me

**Comment by sanposhiho** [comment]
/assign @tangwz 

FYI, @sivchari I reassigned them to @tangwz. You can still help with the review on them.

**Comment by sanposhiho** [comment]
The last one is merged, we can close it now finally 🎉 
https://github.com/kubernetes/kubernetes/pull/117024

/close
